### ISSUE TITLE
Is it possible to dynamically switch between CPU and GPU?

### ISSUE THREAD

**Issue by zhulingchen** [tensorflow/tensorflow#31695]
I have a code that does training/test together.

However, the eager execution mode seems taking more GPU memory than I have when inference is taken place.

I need to train the data in GPU and do inference in CPU. How can I do this in TensorFlow? Thank you!

**Comment by oanush** [comment]
@zhulingchen ,
Please provide details about what platform you are using (operating system, architecture). Also include your TensorFlow version.
In order to expedite the trouble-shooting process, please provide complete code snippet to reproduce the issue reported here. Thanks!

**Comment by zhulingchen** [comment]
I think I solved my problem:

I was intending to use `model(x_test)` to do some inference with TensorFlow Probability `tfp` (with eager execution enabled) since `tfp` does not allow me to use `model.predict(x_test)` yet. But `model(x_test)` reports GPU Resource OOM issue but `model.predict(x_test)` for the regular non-`tfp` model does not have such issue.

So at the very beginning, I wanted to do such inference using `model(x_test)` with CPU instead of GPU to solve the problem. But later, I realized that `model.predict(x_test)` does the inference in a batch-by-batch manner (default batch size is 32) but `model(x_test)` has no such mechanism at all. So finally I designed my manual batch-by-batch inference mechanism for `model(x_test)` and everything now can smoothly work in GPU now. No more GPU Resource OOM issue.
